src/charm.py

```diff
@@ -214,16 +217,44 @@ def _on_start(self, event) -> None:
 
             self.app_peer_data["db_initialised"] = "True"
 
+    def is_unit_in_replica_set(self) -> bool:
+        """Check if the unit is in the replica set."""
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            try:
+                replset_members = mongo.get_replset_members()
+                return self.get_hostname_by_unit(self.unit.name) in replset_members
+            except NotReadyError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set NotReadyError={e}")
+            except PyMongoError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set PyMongoError={e}")
+
+    def _on_stop(self, event) -> None:
+        if "True" == self.unit_peer_data.get("unit_departed", "False"):
+            logger.debug(f"{self.unit.name} blocking on_stop")
+            is_in_replica_set = True
+            timeout = ON_STOP_TIMEOUT
```

Comment:
nit: you could use `tenacity.retry` to abstract the retry logic

Reply:
I don't think that using `renacity` here is the god idea